### PR TITLE
[GenAI] Introduce CausalLMPipelineChatClient for MEAI.IChatClient

### PR DESCRIPTION
**Pull request by LittleLittleCloud**
We are excited to review your PR.

So we can do the best job, please check:

- [x] There's a descriptive title that will make sense to other developers some time from now. 
- [x] There's associated issues. All PR's should have issue(s) associated - unless a trivial self-evident change such as fixing a typo. You can use the format `Fixes #nnnn` in your description to cause GitHub to automatically close the issue(s) when your PR is merged.
- [x] Your change description explains what the change does, why you chose your approach, and anything else that reviewers should know.
- [x] You have included any necessary tests in the same PR.

This PR implements `CausalLMPipelineChatClient<TTokenizer, TCausalLMModel>` class for MEAI.IChatClient's intergration.

This PR also refactors the AutoGen and semantic kernel intergretion for Phi3, which pulls out the prompt template parts into an individual `Phi3ChatTemplateBuilder` class.

This PR also update `STJ` to 8.0.5 to address a security warning for that package

The `CausalLMPipelineChatClient` is used as the abstract class for all `IChatClient` implementation in GenAI models. For now, the following implementations are available
- Phi3CausalLMChatClient
- Llama3CausalLMChatClient

The `MistralCausalLMChatClient` will come in the following PR with additional support for tool call.

## Usage
Checkout the `MEAI` folder under Microsoft.ML.GenAI.Samples` for examples. The following examples are added
- Llama3_1
- Phi3

### CODE REVIEW

File: src/Microsoft.ML.GenAI.Core/CausalLMPipelineChatClient.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    private readonly ICausalLMPipeline<TTokenizer, TCausalLMModel> _pipeline;
+    private readonly ChatClientMetadata _metadata;
+    private readonly IMEAIChatTemplateBuilder _chatTemplateBuilder;
+
+    public CausalLMPipelineChatClient(
+        ICausalLMPipeline<TTokenizer, TCausalLMModel> pipeline,
+        IMEAIChatTemplateBuilder chatTemplateBuilder,
+        ChatClientMetadata? metadata = null)
+    {
+        metadata ??= new ChatClientMetadata(modelId: nameof(CausalLMPipelineChatClient<TTokenizer, TCausalLMModel>));
```
Comment (stephentoub) [inline src/Microsoft.ML.GenAI.Core/CausalLMPipelineChatClient.cs:31]: This makes the modelId "CausalLMPipelineChatClient"... should it actually be `typeof(TCausalLMModel).Name`?

Comment (stephentoub) [reply src/Microsoft.ML.GenAI.Core/CausalLMPipelineChatClient.cs:31]: Also, it'd be nice if the providerName could be set as well.

File: src/Microsoft.ML.GenAI.Core/CausalLMPipelineChatClient.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        _chatTemplateBuilder = chatTemplateBuilder;
+        _pipeline = pipeline;
+        _metadata = metadata;
+    }
+
+    public ChatClientMetadata Metadata => _metadata;
```
Comment (stephentoub) [inline src/Microsoft.ML.GenAI.Core/CausalLMPipelineChatClient.cs:37]: Nit: you can use an auto-prop, e.g. have this be:
```C#
public ChatClientMetadata Metadata { get; }
```
and the in the constructor just do:
```C#
Metadata = metadata ?? new ChatClientMetadata(modelId: nameof(CausalLMPipelineChatClient<TTokenizer, TCausalLMModel>));
```
and delete the _metadata field.

File: src/Microsoft.ML.GenAI.Core/CausalLMPipelineChatClient.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        ChatOptions? options = null,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        var prompt = _chatTemplateBuilder.BuildPrompt(chatMessages, options);
+        var stopSequences = options?.StopSequences ?? Array.Empty<string>();
+
+        foreach (var output in _pipeline.GenerateStreaming(
+            prompt,
+            maxLen: options?.MaxOutputTokens ?? 1024,
+            temperature: options?.Temperature ?? 0.7f,
```
Comment (stephentoub) [inline src/Microsoft.ML.GenAI.Core/CausalLMPipelineChatClient.cs:73]: Does GenerateStreaming require these values to be set? Just curious where these const choices are coming from.

File: src/Microsoft.ML.GenAI.LLaMA/Llama3CausalLMChatClient.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        else
+        {
+            options.StopSequences = new List<string> { _eotToken };
+        }
+
+        return base.CompleteAsync(chatMessages, options, cancellationToken);
+    }
+
+#pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
+    public override async IAsyncEnumerable<StreamingChatCompletionUpdate> CompleteStreamingAsync(
```
Comment (stephentoub) [inline src/Microsoft.ML.GenAI.LLaMA/Llama3CausalLMChatClient.cs:47]:
```suggestion
    public override IAsyncEnumerable<StreamingChatCompletionUpdate> CompleteStreamingAsync(
```

File: src/Microsoft.ML.GenAI.LLaMA/Llama3CausalLMChatClient.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        }
+        else
+        {
+            options.StopSequences = new List<string> { _eotToken };
+        }
+
+        await foreach (var update in base.CompleteStreamingAsync(chatMessages, options, cancellationToken))
+        {
+            yield return update;
+        }
```
Comment (stephentoub) [inline src/Microsoft.ML.GenAI.LLaMA/Llama3CausalLMChatClient.cs:67]:
```suggestion
        return base.CompleteStreamingAsync(chatMessages, options, cancellationToken);
```

File: src/Microsoft.ML.GenAI.LLaMA/Llama3CausalLMChatClient.cs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        options ??= new ChatOptions();
+
+        if (options.StopSequences != null)
+        {
+            options.StopSequences.Add(_eotToken);
+        }
+        else
+        {
+            options.StopSequences = new List<string> { _eotToken };
+        }
```
Comment (stephentoub) [inline src/Microsoft.ML.GenAI.LLaMA/Llama3CausalLMChatClient.cs:62]:
```suggestion
        options.StopSequences ??= [];
        options.StopSequences.Add(_eotToken);
```

File: src/Microsoft.ML.GenAI.LLaMA/Llama3CausalLMChatClient.cs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
```
Comment (stephentoub) [inline src/Microsoft.ML.GenAI.LLaMA/Llama3CausalLMChatClient.cs:41]:
```suggestion
        options.StopSequences ??= [];
        options.StopSequences.Add(_eotToken);
```

File: src/Microsoft.ML.GenAI.Phi/Phi3/Phi3CausalLMChatClient.cs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        options ??= new ChatOptions();
+
+        if (options.StopSequences != null)
+        {
+            options.StopSequences.Add(_eotToken);
+        }
+        else
+        {
+            options.StopSequences = [_eotToken];
+        }
```
Comment (stephentoub) [inline src/Microsoft.ML.GenAI.Phi/Phi3/Phi3CausalLMChatClient.cs:46]: Can be condensed as earlier

File: src/Microsoft.ML.GenAI.Phi/Phi3/Phi3CausalLMChatClient.cs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+        return base.CompleteAsync(chatMessages, options, cancellationToken);
+    }
+
+#pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
+    public override async IAsyncEnumerable<StreamingChatCompletionUpdate> CompleteStreamingAsync(
+#pragma warning restore CS1998 // Async method lacks 'await' operators and will run synchronously
```
Comment (stephentoub) [inline src/Microsoft.ML.GenAI.Phi/Phi3/Phi3CausalLMChatClient.cs:53]:
```suggestion
    public override IAsyncEnumerable<StreamingChatCompletionUpdate> CompleteStreamingAsync(
```

File: src/Microsoft.ML.GenAI.Phi/Phi3/Phi3CausalLMChatClient.cs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        }
+        else
+        {
+            options.StopSequences = new List<string> { _eotToken };
+        }
+
+        await foreach (var update in base.CompleteStreamingAsync(chatMessages, options, cancellationToken))
+        {
+            yield return update;
+        }
```
Comment (stephentoub) [inline src/Microsoft.ML.GenAI.Phi/Phi3/Phi3CausalLMChatClient.cs:72]:
```suggestion
        return base.CompleteStreamingAsync(chatMessages, options, cancellationToken);
```

File: src/Microsoft.ML.GenAI.Phi/Phi3/Phi3ChatTemplateBuilder.cs [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+using Microsoft.SemanticKernel.ChatCompletion;
+using TextContent = Microsoft.SemanticKernel.TextContent;
+
+namespace Microsoft.ML.GenAI.Phi;
+
+public class Phi3ChatTemplateBuilder : IChatTemplateBuilder, IMEAIChatTemplateBuilder
+{
+    private const char Newline = '\n';
+
+    public static Phi3ChatTemplateBuilder Instance => new Phi3ChatTemplateBuilder();
```
Comment (stephentoub) [inline src/Microsoft.ML.GenAI.Phi/Phi3/Phi3ChatTemplateBuilder.cs:23]: Nit: Properties like this are expected to return the same instance on each call. If this is indeed supposed to be creating a new instance on each access, it should be a method.

File: src/Microsoft.ML.GenAI.Phi/Phi3/Phi3ChatTemplateBuilder.cs [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+        var sb = new StringBuilder();
+
+        foreach (var message in chatHistory)
+        {
+            foreach (var item in message.Items)
+            {
+                if (item is not TextContent textContent)
+                {
+                    throw new NotSupportedException($"Only text content is supported, but got {item.GetType().Name}");
+                }
```
Comment (stephentoub) [inline src/Microsoft.ML.GenAI.Phi/Phi3/Phi3ChatTemplateBuilder.cs:73]: In other implementations, we've generally just been ignoring content that's not understood.

File: src/Microsoft.ML.GenAI.Core/CausalLMPipelineChatClient.cs (hunk 1 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.GenAI.Core/CausalLMPipelineChatClient.cs:31]: The provider name should be `CausalLMPipelineChatClient<TTokenizer, TCausalLMModel>`?

Comment (stephentoub) [reply src/Microsoft.ML.GenAI.Core/CausalLMPipelineChatClient.cs:31]:
```C#
Test<int>.Print();

class Test<T>
{
    public static void Print() => Console.WriteLine(nameof(Test<T>));
}
```
prints `Test` not `Test<T>`.

Comment (LittleLittleCloud) [reply src/Microsoft.ML.GenAI.Core/CausalLMPipelineChatClient.cs:31]: Thanks, now I use `$"{nameof(CausalLMPipelineChatClient<TTokenizer, TCausalLMModel>)}<{typeof(TTokenizer).Name}, {typeof(TCausalLMModel).Name}>"` to get the full name with type as model provider.

File: src/Microsoft.ML.GenAI.Phi/Phi3/Phi3ChatTemplateBuilder.cs (hunk 11 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.GenAI.Phi/Phi3/Phi3ChatTemplateBuilder.cs:23]: I've updated it to `public static Phi3ChatTemplateBuilder Instance {get;} => new Phi3ChatTemplateBuilder();`

File: src/Microsoft.ML.GenAI.Phi/Phi3/Phi3ChatTemplateBuilder.cs (hunk 12 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.GenAI.Phi/Phi3/Phi3ChatTemplateBuilder.cs:73]: This template builder is specific to Phi3 and is not expected to be used for other LLMs, so throwing an exception might make more sense here?

File: src/Microsoft.ML.GenAI.Core/CausalLMPipelineChatClient.cs (hunk 3 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.GenAI.Core/CausalLMPipelineChatClient.cs:73]: For `Temperature` yes, it's for sampling strategy. And there needs to be a default value for templerature when sampling even if it's null. `0.7` is just a common figure to choose for sampling. 

`MaxLen` is not though actually so theoratically you can generate as many output tokens as you want, for practical model providers usually set a hard limit to the maximum length of output tokens. Here I set it to 1024 considering that inference speed on local device is ususally much slower than cloud's inference speed. And this figure is also usually large enough for LLM to produce the `stop_token`

File: src/Microsoft.ML.GenAI.Core/Microsoft.ML.GenAI.Core.csproj [hunk 13]
```diff
@@ -13,6 +13,7 @@
 
   <ItemGroup>
     <PackageReference Include="AutoGen.Core" Version="$(AutoGenVersion)" />
+    <PackageReference Include="Microsoft.Extensions.AI" Version="$(MicrosoftExtensionsAIVersion)" />
```
Comment (michaelgsharp) [inline src/Microsoft.ML.GenAI.Core/Microsoft.ML.GenAI.Core.csproj:16]: So is the Extensions.AI package going to be a new permanent dependency for this package?

Comment (michaelgsharp) [reply src/Microsoft.ML.GenAI.Core/Microsoft.ML.GenAI.Core.csproj:16]: Or would it be better to try and make it like an addon package for your gen ai stuff?

Comment (stephentoub) [reply src/Microsoft.ML.GenAI.Core/Microsoft.ML.GenAI.Core.csproj:16]: This should depend on the M.E.AI.Abstractions assembly rather than M.E.AI.

**Review by michaelgsharp — APPROVED** [review]

**Pull request merged**